File: 744. Find Smallest Letter Greater Than Target/binary-search.cpp
// Time : O(n)
// Space : O(1)

// Desc : lower bound for p(x) is true.

class Solution {
private:
    // predicate : letters[mid] > target
    bool p(vector<char>& letters, int mid, char target) {
        return letters[mid] > target;
    }
    
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int lo = 0;
        int hi = letters.size() - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (p(letters, mid, target)) {
                hi = mid;
            }
            else {
                lo = mid + 1;
            }
        }
        
        if (!p(letters, lo, target)) return letters[0];
        return letters[lo];
    }
};

File: 744. Find Smallest Letter Greater Than Target/binary-search.py
class Solution(object):
    def nextGreatestLetter(self, letters, target):
        """
        :type letters: List[str]
        :type target: str
        :rtype: str
        find a first x for which letters[x] > target.
        """
        
        lo = 0
        hi = len(letters) - 1
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if letters[mid] > target:
                hi = mid
            else:
                lo = mid + 1
        if not letters[lo] > target:
            return letters[0]
        return letters[lo]